tests/50federation/52soft-fail.pl

```diff
@@ -323,3 +323,214 @@
          );
       });
    };
+
+
+test "Inbound federation correctly handles soft failed events as extremities",
+   # this is mostly a regression test for https://github.com/matrix-org/synapse/issues/5269.
+   requires => [
+      $main::OUTBOUND_CLIENT, $main::INBOUND_SERVER, $main::HOMESERVER_INFO[0],
+      local_user_and_room_fixtures(
+         user_opts => { with_events => 1 },
+      ),
+      federation_user_id_fixture(),
+   ],
+
+   do => sub {
+      my (
+         $outbound_client, $inbound_server, $info, $creator, $room_id,
+         $remote_user_id,
+      ) = @_;
+      my $first_home_server = $info->server_name;
+
+      my $local_server_name = $outbound_client->server_name;
+
+      my $room;
+
+      # We're going to construct a room graph like:
+      #
+      #        J1
+      #       /  \
+      #      /    \
+      #    PL1     M1
+      #     |      |
+      #     |     SF2
+      #     |      |
+      #      \    SF1
+      #       \   /
+      #         M2
+      #         |
+      #         M3
+      #
+      # Where time is flowing downwards.
+      #
+      #  J1  = join of remote user
+      #  PL1 = creator of room blocks SF event sending
+      #  M1  = remote user sends a permitted message
+      #  SF1 = remote user sends a soft-failed message
+      #  SF2 = remote user sends a soft-failed message
```

Comment:
I think SF2 is unnecessary in this test? you could get rid of it, if you could be bothered.

Reply:
Yes, ish, though it validates the logic a bit better since the test then checks the correct handling of chains of soft failed events.